File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/data/repository/CmpRepositoryImpl.java
package com.bits.bee.bpmc.data.repository;

import java.lang.System;

/**
 * Created by aldi on 20/05/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000,\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\u0018\u00002\u00020\u0001B\u001f\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u00a2\u0006\u0002\u0010\bJ\u0014\u0010\t\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\f0\u000b0\nH\u0016R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u0006\r"}, d2 = {"Lcom/bits/bee/bpmc/data/repository/CmpRepositoryImpl;", "Lcom/bits/bee/bpmc/domain/repository/CmpRepository;", "defaultDispatcher", "Lkotlinx/coroutines/CoroutineDispatcher;", "cmpDao", "Lcom/bits/bee/bpmc/data/data_source/local/dao/CmpDao;", "apiUtils", "Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;", "(Lkotlinx/coroutines/CoroutineDispatcher;Lcom/bits/bee/bpmc/data/data_source/local/dao/CmpDao;Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;)V", "getLatestCmp", "Lkotlinx/coroutines/flow/Flow;", "Lcom/bits/bee/bpmc/utils/Resource;", "Lcom/bits/bee/bpmc/domain/model/Cmp;", "app_debug"})
public final class CmpRepositoryImpl implements com.bits.bee.bpmc.domain.repository.CmpRepository {
    private final kotlinx.coroutines.CoroutineDispatcher defaultDispatcher = null;
    private final com.bits.bee.bpmc.data.data_source.local.dao.CmpDao cmpDao = null;
    private final com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils = null;
    
    @javax.inject.Inject()
    public CmpRepositoryImpl(@org.jetbrains.annotations.NotNull()
    kotlinx.coroutines.CoroutineDispatcher defaultDispatcher, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.local.dao.CmpDao cmpDao, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<com.bits.bee.bpmc.domain.model.Cmp>> getLatestCmp() {
        return null;
    }
}

File: app/src/main/java/com/bits/bee/bpmc/data/repository/CmpRepositoryImpl.kt
package com.bits.bee.bpmc.data.repository

import com.bits.bee.bpmc.data.data_source.local.dao.CmpDao
import com.bits.bee.bpmc.data.data_source.remote.ApiUtils
import com.bits.bee.bpmc.data.data_source.remote.response.CmpResponse
import com.bits.bee.bpmc.domain.mapper.CmpDataMapper
import com.bits.bee.bpmc.domain.model.Cmp
import com.bits.bee.bpmc.domain.repository.CmpRepository
import com.bits.bee.bpmc.utils.ApiResponse
import com.bits.bee.bpmc.utils.NetworkDatabaseBoundResource
import com.bits.bee.bpmc.utils.Resource
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import javax.inject.Inject

/**
 * Created by aldi on 20/05/22.
 */
class CmpRepositoryImpl @Inject constructor(
    private val defaultDispatcher: CoroutineDispatcher,
    private val cmpDao: CmpDao,
    private val apiUtils: ApiUtils
) : CmpRepository {

    override fun getLatestCmp(): Flow<Resource<Cmp>> {
        return object : NetworkDatabaseBoundResource<Cmp, CmpResponse>() {

            override suspend fun loadFormDB(): Cmp? {
                return CmpDataMapper.fromDbToDomain(cmpDao.getCmp())
            }

            override fun shouldFetch(data: Cmp?): Boolean {
                return true
            }

            override suspend fun createCall(): Flow<ApiResponse<CmpResponse>> {
                return apiUtils.getCmpApiService().getCmp()
            }

            override suspend fun saveCallResult(data: CmpResponse) {
                cmpDao.insertSingle(CmpDataMapper.fromNetworkToDb(data.data))
            }
        }.getAsFlow()
    }

    override fun getCmp(): Flow<Cmp> {
        return flow {
            val data = cmpDao.getCmp()
            emit(CmpDataMapper.fromDbToDomain(data))
        }.flowOn(defaultDispatcher)
    }

}